beaker/logic_signature.py

```diff
@@ -1,28 +1,117 @@
-from inspect import getattr_static
-from typing import Optional
-from algosdk.v2client.algod import AlgodClient
+import inspect
+from typing import Callable
+
 from pyteal import (
     CompileOptions,
-    TealInputError,
     TealType,
     Tmpl,
     Expr,
-    MAX_TEAL_VERSION,
-    SubroutineDefinition,
-    ABIReturnSubroutine,
     Seq,
     compileTeal,
-    Reject,
     Mode,
     ScratchVar,
     TealBlock,
     TealSimpleBlock,
 )
-from beaker.decorators import get_handler_config
-from beaker.precompile import AppPrecompile, LSigPrecompile
 
+from beaker.build_options import BuildOptions
+
+__all__ = [
+    "LogicSignature",
+    "LogicSignatureTemplate",
+]
+
+
+class LogicSignature:
+    """
+    LogicSignature allows the definition of a logic signature program.
+
+    A LogicSignature may include constants, subroutines, and :ref:TemplateVariables as attributes
+
+    The `evaluate` method is the entry point to the application and must be overridden in any subclass
+    to call the necessary logic.
+    """
+
+    def __init__(
+        self,
+        expr_or_func: Callable[[], Expr] | Expr,
+        *,
+        build_options: BuildOptions | None = None,
+    ):
+        logic: Expr
+        if callable(expr_or_func):
+            logic = expr_or_func()
+        else:
+            logic = expr_or_func
+
+        self.program = _lsig_to_teal(logic, build_options=build_options)
 
-class TemplateVariable(Expr):
+
+class LogicSignatureTemplate:
+    """
+    LogicSignature allows the definition of a logic signature program.
+
+    A LogicSignature may include constants, subroutines, and :ref:TemplateVariables as attributes
+
+    The `evaluate` method is the entry point to the application and must be overridden in any subclass
+    to call the necessary logic.
+    """
+
+    def __init__(
+        self,
+        expr_or_func: Callable[..., Expr] | Expr,
+        *,
+        runtime_template_variables: dict[str, TealType],
+        build_options: BuildOptions | None = None,
+    ):
+        """initialize the logic signature and identify relevant attributes"""
+        if not runtime_template_variables:
+            raise ValueError(
+                "No runtime template variables supplied - use LogicSignature instead if that was intentional"
+            )
+
+        build_options = build_options or BuildOptions()
+
+        self.runtime_template_variables: dict[str, RuntimeTemplateVariable] = {
+            name: RuntimeTemplateVariable(stack_type=stack_type, name=name)
+            for name, stack_type in runtime_template_variables.items()
+        }
+
+        logic: Expr
+        if not callable(expr_or_func):
+            logic = expr_or_func
+        else:
+            params = inspect.signature(expr_or_func).parameters
+            if not (params.keys() <= runtime_template_variables.keys()):
```

Comment:
should this be `len(params.keys()) <= ...`?

Reply:
i think yes, but also TIL ```
(.venv) ben@Algo-Guidarelli:~/beaker/examples/templated_lsig$ python
Python 3.10.4 (main, Mar 24 2022, 16:12:56) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> a = {"a":"c"}
>>> b = {"b":"d"}
>>> ak = a.keys()
>>> bk = b.keys()
>>> ak 
dict_keys(['a'])
>>> ak >= bk
False
>>> ak <= bk
False
>>> ak == bk
False
>>> b = {"a":"d"}
>>> bk = b.keys()
>>> ak == bk
True
>>> ak >= bk
True
>>> ak <= bk
True
```

Reply:
This way it will check that the arguments the function take are correctly named, which a length check would not

Reply:
ah I see now, its doing a set operation.


```
>>> a = {"n":"m"}
>>> b = {"a":"c", "n":"m"}
>>> ak = a.keys()
>>> bk = b.keys()
>>> ak <= bk
True
>>> ak >= bk
False
>>> ak == bk
False
>>> ak & bk
{'n'}
>>> bk | bk
{'a', 'n'}
>>> ak ^ bk
{'a'}
```

In that case, we should probably modify the error message to spit out the expected template vars vs actual since if the names dont match it will throw a confusing error message

Reply:
Something like this?

```py
           # ...
            param_keys = inspect.signature(expr_or_func).parameters.keys()
            template_keys = runtime_template_variables.keys()
            if len(param_keys) == 0 and len(template_keys)>0:
                raise ValueError(
                    "Logic signature methods should take no arguments, "
                    "unless using runtime templates"
                )

            if not (param_keys <= template_keys):
                raise ValueError(
                    f"Logic signature expected ({','.join(param_keys)}), "
                    f"got ({','.join(template_keys)})"
                )

            forward_args = list(param_keys)
            # ...
```

The `templated_lsig` example can be used to see the result for this by changing the key in the `runtime_template_variables` argument

Reply:
Oh wow yes that was a terrible error message I had there. Lots of them need attention actually, hopefully none so utterly wrong as this one was. I'll add a todo checklist item for checking error message strings.

I've changed the error message, although it's slightly different to your suggestion. The first one isn't quite right - template keys will always be non-empty at that point (there's an earlier check already:

```python3
class LogicSignatureTemplate:
    """
    LogicSignature allows the definition of a logic signature program.

    A LogicSignature may include constants, subroutines, and :ref:TemplateVariables as attributes

    The `evaluate` method is the entry point to the application and must be overridden in any subclass
    to call the necessary logic.
    """

    def __init__(
        self,
        expr_or_func: Callable[..., Expr] | Expr,
        *,
        runtime_template_variables: dict[str, TealType],
        build_options: BuildOptions | None = None,
    ):
        """initialize the logic signature and identify relevant attributes"""
        if not runtime_template_variables:
            raise ValueError(
                "No runtime template variables supplied - use LogicSignature instead if that was intentional"
            )
```

Also, it's totally valid to not take any arguments - `examples/account_storage` is one such example.

For the second error message, I've made it so it will just print out the invalid argument names.